Lib/LVGL9/callback/callback.c

```diff
@@ -0,0 +1,59 @@
+/**
+ * @file    callback.c
+ * @author  Matteo Giuliani [matteo.giuliani.sc@gmail.com]
+ * @date    2024-12-06 (date of creation)
+ * @version v0.0.1
+ * @prefix  TMP
+ *
+ * @brief   Implementation of callback
+ *
+ * @license Licensed under "THE BEER-WARE LICENSE", Revision 69 
+ *          see LICENSE file in the root directory of this software component
+ */
+
+/*---------- Includes -------------------------------------------------------*/
+
+#include "callback.h"
+
+/*---------- Private define -------------------------------------------------*/
+
+/*---------- Private macro --------------------------------------------------*/
+
+/*---------- Private variables ----------------------------------------------*/
+uint8_t buf1[BUFFER_SIZE];
+static volatile uint8_t flush_in_progress = 0;
+
+/*---------- Private function prototypes ------------------------------------*/
+
+/*---------- Exported Variables ---------------------------------------------*/
+
+/*---------- Exported Functions ---------------------------------------------*/
+
+/*---------- Private Functions ---------------------------------------------*/
+void lvgl_flush_callback(lv_display_t *display, const lv_area_t *area, uint8_t *px_map) {
+    if (flush_in_progress) {
+        return;
+    }
+
+    flush_in_progress = 1;  // Imposta il flag per indicare che la trasmissione è in corso
+
+    set_draw_window(area->x1, area->y1, area->x2, area->y2);
+
+    uint32_t width       = area->x2 - area->x1 + 1;
+    uint32_t height      = area->y2 - area->y1 + 1;
+    uint32_t pixel_count = width * height;
+
+    HAL_GPIO_WritePin(LCD_TFT_DC_GPIO_Port, LCD_TFT_DC_Pin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(LCD_TFT_CS_GPIO_Port, LCD_TFT_CS_Pin, GPIO_PIN_RESET);
+
+    HAL_SPI_Transmit_DMA(&hspi3, px_map, pixel_count * 3);
+}
+
+void HAL_SPI_TxCpltCallback(SPI_HandleTypeDef *hspi) {
+    if (hspi->Instance == SPI3) {
+        HAL_GPIO_WritePin(LCD_TFT_CS_GPIO_Port, LCD_TFT_CS_Pin, GPIO_PIN_SET);
```

Comment:
Take a look at bsp.h and read and understand how to use the struct GPIO_Tuple. Then you can define a section for the LCD_TFT in the bsp.c/h files and put all this stuf, e.g:
```c
//In bsp.h
/* LCD_TFT #################################################/
#define LCD_TFT_SPI_HANDLE hspi3
void LCD_TFT_CS_setState(uint8_t state);
// In bsp.c
/* LCD_TFT #################################################/
static struct GPIO_Tuple LCD_TFT_CS {.GPIO_Port = LCD_TFT_CS_GPIO_Port, .GPIO_Pin = LCD_TFT_CS_Pin};
void LCD_TFT_CS_setState(uint8_t state){
    HAL_GPIO_WritePin(LCD_TFT_CS.Gpio_Port, LCD_TFT_CS.GPIO_Pin,state);
};
```

Reply:
Why would we use a wrapper function to toggle the GPIO?
Advantages:
- if someday the pin has an inverted logic on the pcb you change the call inside the wrapper function and is reflected everywhere instead of searching manually for all references  

Disadvantages:
- additional context switch for entering the function -> speed penalty -> but usually neglegible

I suggest you reeding the bsp.c/h files and the functions in it and how i use them. Then add new peripherals to it following the same approach. Remember the talk I gave months ago on abstraction